main.rb

```diff
@@ -0,0 +1,122 @@
+# Enumerable
+module Enumerable
+  # rubocop:disable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
+
+  def my_each
+    return to_enum(__method__) unless block_given?
+
+    if is_a?(Array)
+      size.times do |i|
+        yield self[i]
+      end
+    else size.times do |i|
+      yield [keys[i], values[i]]
+    end
+    end
+  end
+
+  def my_each_with_index
+    return to_enum(__method__) unless block_given?
+
+    size.times do |i|
+      yield self[i], i
+    end
+  end
+
+  def my_select
+    return to_enum(__method__) unless block_given?
+
+    array = []
+    my_each do |i|
+      array << i if yield i
+    end
+    array
+  end
+
+  def my_all?(*args)
+    if block_given?
+      my_each { |i| return false unless yield i }
+    elsif args != [] && args[0].class != Regexp
+      my_each { |i| return false unless i.is_a?(args[0]) }
+    elsif args[0].class == Regexp
+      my_each { |i| return false if (args[0] =~ i).nil? }
+    else
+      my_each { |i| return false unless i.nil? }
+    end
+    true
+  end
+
+  def my_any?(*args)
+    if block_given?
+      my_each { |i| return true if yield i }
+    elsif args != [] && args[0].class != Regexp
+      my_each { |i| return true if i.is_a?(args[0]) }
+    elsif args[0].class == Regexp
+      my_each { |i| return true unless (args[0] =~ i).nil? }
+    else
+      my_each { |i| return true if i.nil? }
+    end
+    false
+  end
+
+  def my_none?(*args)
+    if block_given?
+      my_each { |i| return false if yield i }
+    elsif args != [] && args[0].class != Regexp
+      my_each { |i| return false if i.is_a?(args[0]) }
+    elsif args[0].class == Regexp
+      my_each { |i| return false unless (args[0] =~ i).nil? }
+    else
+      my_each { |i| return false if i == true }
+    end
+    true
+  end
+
+  def my_count(*args)
+    @var = self
+    empty_arr = []
+
+    if block_given?
+      my_each { |i| empty_arr << i if yield i }
+
+      empty_arr.length
+    else
+      !args.empty? ? @var.length - (self - args).length : length
+    end
+  end
+
+  def my_map(*args)
+    return to_enum(__method__) unless block_given? || args[0].class == Proc
+
+    array = []
+    args[0].class == Proc ? my_each { |i| array << args[0].call(i) } : my_each { |i| array << (yield i) }
+    array
+  end
+
+  def my_inject(*args)
+    accu = 1
+    sym = args.length > 1 ? args[1] : args[0]
+    if (args.length > 1) && (args[0].object_id != :*.object_id)
+      accu = args[0]
+    elsif (args.length == 1) && (args[0].class != Symbol)
+      accu = args[0]
+    elsif args[0].object_id != :*.object_id
+      accu = 0
+    end
+    if block_given? && to_a[0].is_a?(String)
+      accu = self
+      sym = nil
+      my_each { |i| accu = yield accu, i }
+    elsif block_given?
+      my_each { |i| accu = yield accu, i }
+    else
+      my_each { |i| accu = accu.method(sym).call(i) }
+    end
+    accu
+  end
```

Comment:
- [ ] should have working `my_inject` **(covered by automated tests)** identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [ ] when a block is given without an initail value combines all elements of enum by applying a binary operation, specified by a block::range
```ruby
(1..5).my_inject { |prod, n| prod * n } == (1..5).inject { |prod, n| prod * n } => 120
```
- Yours is throwing an error
![image](https://user-images.githubusercontent.com/52209067/85832209-e1af9700-b74c-11ea-9398-de79643bc2d7.png)

- Real inject
![image](https://user-images.githubusercontent.com/52209067/85832241-f1c77680-b74c-11ea-8282-1216da76fe15.png)

- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```
- Yours is retrieving an error
![image](https://user-images.githubusercontent.com/52209067/85831980-76fe5b80-b74c-11ea-9232-3f8b815fb827.png)

- Real inject
![image](https://user-images.githubusercontent.com/52209067/77698409-2ffaa180-6f76-11ea-91d3-9778aa560411.png)

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
(1..5).my_inject(:+) == (1..5).inject(:+) => 15
```
- Yours is retrieving an Error
![image](https://user-images.githubusercontent.com/52209067/85832377-34894e80-b74d-11ea-8842-47c36bb24fec.png)

- Real inject
![image](https://user-images.githubusercontent.com/52209067/85832402-42d76a80-b74d-11ea-9b95-5651b5ccef98.png)

- [ ] when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

```ruby
(1..5).my_inject(2, :*) == (1..5).inject(2, :*) => 240
```
- Yours is retrieving an error
![image](https://user-images.githubusercontent.com/52209067/85832592-9d70c680-b74d-11ea-88b0-f56ea96929d0.png)

- Real inject
![image](https://user-images.githubusercontent.com/52209067/85832634-b11c2d00-b74d-11ea-92a6-19b675add319.png)

Reply:
Thank you very much for your review euqueme, it was super detailed and helped us a lot!